server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java

```diff
@@ -0,0 +1,230 @@
+        return this.scriptScoreFunctionBuilder;
+    }
+
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (query != null) {
+            builder.field(QUERY_FIELD.getPreferredName());
+            query.toXContent(builder, params);
+        }
+        builder.field(SCRIPT_FIELD.getPreferredName(), scriptScoreFunctionBuilder.getScript());
+        if (minScore != null) {
+            builder.field(MIN_SCORE_FIELD.getPreferredName(), minScore);
+        }
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    public ScriptScoreQueryBuilder setMinScore(float minScore) {
+        this.minScore = minScore;
+        return this;
+    }
+
+    public Float getMinScore() {
+        return this.minScore;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected boolean doEquals(ScriptScoreQueryBuilder other) {
+        return Objects.equals(this.query, other.query) &&
+            Objects.equals(this.scriptScoreFunctionBuilder, other.scriptScoreFunctionBuilder) &&
+            Objects.equals(this.minScore, other.minScore) ;
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(this.query, this.scriptScoreFunctionBuilder, this.minScore);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        ScriptScoreFunction function = (ScriptScoreFunction) scriptScoreFunctionBuilder.toFunction(context);
+        Query query = this.query.toQuery(context);
+        if (query == null) {
+            query = new MatchAllDocsQuery();
+        }
+        return new ScriptScoreQuery(query, function, minScore);
+    }
+
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder newQuery = this.query.rewrite(queryRewriteContext);
+        if (newQuery != query) {
+            ScriptScoreQueryBuilder newQueryBuilder = new ScriptScoreQueryBuilder(newQuery, scriptScoreFunctionBuilder);
+            newQueryBuilder.setMinScore(minScore);
+            return newQueryBuilder;
+        }
+        return this;
+    }
+
+    @Override
+    protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
+        InnerHitContextBuilder.extractInnerHits(query(), innerHits);
+    }
+
+    public static ScriptScoreQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        QueryBuilder query = null;
+        ScriptScoreFunctionBuilder functionBuilder = null;
+        String queryName = null;
+        Float minScore = null;
+        float boost = AbstractQueryBuilder.DEFAULT_BOOST;
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
Is there a reason this parsing can't be done using either `ObjectParser` or `ConstructingObjectParser`? I think if we can do it with one of those it would be great as its easier to see what they are doing